File: backend/src/routes/customers.js
const express = require("express");
const router = express.Router();
const pool = require("../db");

// GET /api/customers?search=
router.get("/", async (req, res) => {
  try {
    const { search } = req.query;
    let query = "SELECT * FROM customers";
    const params = [];

    if (search) {
      query += " WHERE name ILIKE $1 OR phone ILIKE $1";
      params.push(`%${search}%`);
    }
    query += " ORDER BY id DESC";

    const result = await pool.query(query, params);
    res.json(result.rows);
  } catch (err) {
    console.error(err);
    res.status(500).json({ error: "Failed to fetch customers" });
  }
});

// GET /api/customers/:id
router.get("/:id", async (req, res) => {
  try {
    const { id } = req.params;
    const result = await pool.query("SELECT * FROM customers WHERE id = $1", [id]);
    if (result.rows.length === 0) {
      return res.status(404).json({ error: "Customer not found" });
    }
    res.json(result.rows[0]);
  } catch (err) {
    console.error(err);
    res.status(500).json({ error: "Failed to fetch customer" });
  }
});

// POST /api/customers
router.post("/", async (req, res) => {
  try {
    const { name, phone, address } = req.body;
    if (!name || !phone) {
      return res.status(400).json({ error: "Name and phone are required" });
    }
    const result = await pool.query(
      "INSERT INTO customers (name, phone, address) VALUES ($1, $2, $3) RETURNING *",
      [name, phone, address || null]
    );
    res.status(201).json(result.rows[0]);
  } catch (err) {
    console.error(err);
    if (err.code === "23505") {
      return res.status(409).json({ error: "Phone number already exists" });
    }
    res.status(500).json({ error: "Failed to create customer" });
  }
});

// PUT /api/customers/:id
router.put("/:id", async (req, res) => {
  try {
    const { id } = req.params;
    const { name, phone, address } = req.body;
    if (!name || !phone) {
      return res.status(400).json({ error: "Name and phone are required" });
    }
    const result = await pool.query(
      "UPDATE customers SET name = $1, phone = $2, address = $3, updated_at = NOW() WHERE id = $4 RETURNING *",
      [name, phone, address || null, id]
    );
    if (result.rows.length === 0) {
      return res.status(404).json({ error: "Customer not found" });
    }
    res.json(result.rows[0]);
  } catch (err) {
    console.error(err);
    if (err.code === "23505") {
      return res.status(409).json({ error: "Phone number already exists" });
    }
    res.status(500).json({ error: "Failed to update customer" });
  }
});

// DELETE /api/customers/:id
router.delete("/:id", async (req, res) => {
  try {
    const { id } = req.params;
    const result = await pool.query("DELETE FROM customers WHERE id = $1 RETURNING *", [id]);
    if (result.rows.length === 0) {
      return res.status(404).json({ error: "Customer not found" });
    }
    res.json({ message: "Customer deleted" });
  } catch (err) {
    console.error(err);
    res.status(500).json({ error: "Failed to delete customer" });
  }
});

module.exports = router;

File: backend/app/routers/customers.py
from fastapi import APIRouter, Depends, HTTPException, Query, status
from sqlalchemy.orm import Session
from typing import Optional

from ..database import get_db
from ..models import Customers
from ..schemas import CustomerCreate, CustomerUpdate, CustomerOut, MessageResponse
from ..auth import get_current_user

router = APIRouter(prefix="/api/customers", tags=["customers"])


@router.get("/", response_model=list[CustomerOut])
def list_customers(
    search: Optional[str] = Query(None),
    db: Session = Depends(get_db),
    user_id: int = Depends(get_current_user),
):
    q = db.query(Customers)
    if search:
        like = f"%{search}%"
        q = q.filter((Customers.name.ilike(like)) | (Customers.phone.ilike(like)))
    return q.order_by(Customers.id.desc()).all()


@router.get("/{id}", response_model=CustomerOut)
def get_customer(
    id: int,
    db: Session = Depends(get_db),
    user_id: int = Depends(get_current_user),
):
    customer = db.query(Customers).filter(Customers.id == id).first()
    if not customer:
        raise HTTPException(status_code=404, detail="Customer not found")
    return customer


@router.post("/", response_model=CustomerOut, status_code=201)
def create_customer(
    body: CustomerCreate,
    db: Session = Depends(get_db),
    user_id: int = Depends(get_current_user),
):
    existing = db.query(Customers).filter(Customers.phone == body.phone).first()
    if existing:
        raise HTTPException(status_code=409, detail="Phone number already exists")
    customer = Customers(**body.model_dump())
    db.add(customer)
    db.commit()
    db.refresh(customer)
    return customer


@router.put("/{id}", response_model=CustomerOut)
def update_customer(
    id: int,
    body: CustomerUpdate,
    db: Session = Depends(get_db),
    user_id: int = Depends(get_current_user),
):
    customer = db.query(Customers).filter(Customers.id == id).first()
    if not customer:
        raise HTTPException(status_code=404, detail="Customer not found")

    # Check phone uniqueness if changed
    if body.phone != customer.phone:
        existing = (
            db.query(Customers)
            .filter(Customers.phone == body.phone, Customers.id != id)
            .first()
        )
        if existing:
            raise HTTPException(status_code=409, detail="Phone number already exists")

    for key, value in body.model_dump().items():
        setattr(customer, key, value)
    db.commit()
    db.refresh(customer)
    return customer


@router.delete("/{id}", response_model=MessageResponse)
def delete_customer(
    id: int,
    db: Session = Depends(get_db),
    user_id: int = Depends(get_current_user),
):
    customer = db.query(Customers).filter(Customers.id == id).first()
    if not customer:
        raise HTTPException(status_code=404, detail="Customer not found")
    db.delete(customer)
    db.commit()
    return {"message": "Customer deleted"}
